[Pranav Ganesh] iP

pranav-ganesh opened the pull request:
### :fire:  DUKE :fire: 

> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

Duke frees your mind of having to remember things you need to do. It's, 

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,
1. download it from [here.](https://github.com/pranav-ganesh/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks
4. let it manage your tasks for you :smiley: :blush:

And it is **FREE!**

Duke's features:

- [x] Managing features
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
#

If you Java programmer, you can use it to practice Java too. Here's the  ``` main``` method:
```java
public static void main(String[] args) {
    new Duke("./duke.txt").run();
}
```

Commit:
Add Gradle support

e0661424 committed:
Finished Level-1

e0661424 committed:
Finished Level-2

e0661424 committed:
Finished Level-3

e0661424 committed:
Finished Level-4

e0661424 committed:
Finished A-TextUiTesting

e0661424 committed:
Finished Level-5

e0661424 committed:
Finished Level-6

e0661424 committed:
Comply with Java coding standard

e0661424 committed:
Finish Level-7

e0661424 committed:
Finish Level-8

e0661424 committed:
Add line wrapping at appropriate places

e0661424 committed:
Merge branch 'branch-Level-7'
Merge branch branch-Level-7 into Master

e0661424 committed:
Merge branch branch-Level-8 into Master

e0661424 committed:
Add new files for individual classes in Duke.java

e0661424 committed:
Add TaskList class

e0661424 committed:
Add Storage class

e0661424 committed:
Add Command class and relevant subclasses

e0661424 committed:
Enhance the functionality of the Storage class

e0661424 committed:
Add Ui class

e0661424 committed:
Finish A-MoreOOP

e0661424 committed:
Add A-Packages

e0661424 committed:
Add JUnit Test for the Parser class

e0661424 committed:
Add JUnit Test for the DukeException class

e0661424 committed:
Add JUnit Test for the TaskList class

e0661424 committed:
Finish A-JUnit

yongxiangng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+package duke;
+
+import duke.command.*;
```
You might want to avoid wild card imports even if you have to import everything 
https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements

yongxiangng replied:
For the `parse` method below

Perhaps a switch statement here would help to improve readability or you could modify the code by splitting the string by the empty character delimiter twice so you don't have to do another `if` inside the external `if`

Maybe your InvalidCommand can be an exception, and your doneCommand can throw the exception so you don't have to do all the checking within this method which could make the code easier to read

https://www.geeksforgeeks.org/split-string-java-examples/

https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#guideline-maximise-readability

yongxiangng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+    
+    public void loadTaskListData(TaskList taskList) {
```
You may want to document public methods with javadocs
https://se-education.org/guides/conventions/java/intermediate.html#comments

The method here is also quite lengthy. It might be a good idea to abstract the other logic into various methods
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods

Linxcathyyy replied:
You might want to break it down into smaller functions here? This function seems to be too deeply nested.

yongxiangng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+                    System.out.println(str);
+                    String[] parts = str.split("\\|", 4);
+                    String subStr = parts[0].substring(3).trim();
+                    if (subStr.equals("duke.task.Todo") || subStr.equals("duke.task.Todo")) {
```
I might be wrong here but doesn't `subStr.equals("duke.task.Todo") || subStr.equals("duke.task.Todo")` do the same as `subStr.equals("duke.task.Todo")`

yongxiangng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+    public void loadTaskListData(TaskList taskList) {
+        try {
+            Scanner s = new Scanner(this.file); // create a Scanner using the File as the source
+            if (!s.hasNext()) {
```
The logic here is quite deeply nested, perhaps it might be good if you handled the unusual case with a guard clause
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent
```java
if (!s.hasNext()) {
    // code here...
    return;
}
```

yongxiangng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,84 @@
+                            task.markAsDone();
+                        }
+                        taskList.addTask(task);
+                    } else {
```
You may want to use a switch statement for this long if else. It may be good because you're comparing strings and it can directly be used for your cases.

Additionally the else statement here captures all other cases of subStr, not necessarily when it is an `Event`. If new requirements come in and there are new type of `Task`, the developer might forget to update the logic here and the new type of `Task` will automatically fall into the else case and treated like an `Event`.

The default case in a switch statement could be used to catch invalid cases. I believe the default case is optional in Java
https://www.geeksforgeeks.org/switch-statement-in-java/

The style guide suggests that
```
Always include a default branch in case statements.

Furthermore, use it for the intended default action and not just to execute the last option. If there is no default action, you can use the default branch to detect errors (i.e. if execution reached the default branch, raise a suitable error). This also applies to the final else of an if-else construct. That is, the final else should mean 'everything else', and not the final option. Do not use else when an if condition can be explicitly specified, unless there is absolutely no other possibility.
```

yongxiangng reviewed (commented):
There are many Javadocs missing, it would be good to write header comments for all public classes and methods see [here](https://se-education.org/guides/conventions/java/intermediate.html#comments)

It might also be good to be consistent with the number of new lines at the end of your files rather than no or dangling new lines. I believe it is good to have a single new line at the end.

There are several places where `System.out.println` was directly used, possibly violating the abstraction from `Ui.printMessage` (I guess this might not really matter once you start implementing the GUI, you wouldn't be able to use `System.out.println` to directly print to your UI)

Linxcathyyy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+        this.command = command;
+    }
+    
+    public Command parse() {
```
Perhaps try not to write deeply nested loops in your function? You might consider using 'switch' for this function?☺️

Linxcathyyy reviewed (commented):
Good job overall! Just one thing to take note of: try to use small and simple functions to handle your complex logic, don't put everything into one function. It makes readers confusing and it is also hard to do unit testing~

e0661424 committed:
Modify .gitignore file

SKCheongBrian commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,60 @@
+                && command.length() <= 8 && !Character.isAlphabetic(command.charAt(command.length() - 2))
+                && Character.isDigit(command.charAt(5))) {
+            return new doneCommand(command);
+        } else {
+            if (command.startsWith("todo")) {
+                return new todoCommand(command);
+            } else if (command.startsWith("deadline")) {
+                return new deadlineCommand(command);
+            } else if (command.startsWith("event")) {
+                return new eventCommand(command);
+            } else if (command.startsWith("delete")) {
+                return new deleteCommand(command);
+            } else {
+                return new InvalidCommand(command);    
```
You can consider using a switch statement here 😄

SKCheongBrian commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,74 @@
+        System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+    }
+    
+    public static void doneResponse(Task task) {
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println(task);
+        storage.writeToFile("./duke.txt", taskList);
+    }
```
Good idea to write to file everytime a task is edited. This way the state of the tasks are not lost incase the app crashes.

SKCheongBrian commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,37 @@
+        boolean isExit = false;
+        while (!isExit) {
+            String command = ui.command();
+            Parser parser = new Parser(command);
+            if (parser.isExit()) {
+                break;
+            }
+            Command c = parser.parse();
```
You can consider making `parse()` a static method, that way you would not have to create a new parser object everytime a command is given.

SKCheongBrian reviewed (commented):
Good job! Your code seems very well done.

e0661424 committed:
Finish A-CodingStandard

e0661424 committed:
Finish A-JavaDoc

e0661424 committed:
Finish Level-9

e0661424 committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc into Master

e0661424 committed:
Merge branch-A-CodingStandard into Master

e0661424 committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9 into Master

e0661424 committed:
Finish Level-9 (Latest)

e0661424 committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Merge add-gradle-support into Master

e0661424 committed:
Modify project directory structure

Smanmos commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,69 @@
+            return new listCommand(command);
+        } else if (command.startsWith("done") && Character.isDigit(command.charAt(command.length() - 1)) && command.length() <= 8 
+                && !Character.isAlphabetic(command.charAt(command.length() - 2))  && Character.isDigit(command.charAt(5))) {
+            return new doneCommand(command);
```
This if block's condition could be simplified, with only the command type being checked in the main condition, and the other checks could be further done inside the body.

Smanmos commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * Checks if the user types the command 'bye' or the user clicks enter without typing any command.
+     *
+     * @return True or False.
```
The return value could be further elaborated.

Smanmos commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,69 @@
+ * Deals with making sense of the user command
+ */
+public class Parser {
+    private String command;
```
The string name `command` can be confusing, because a class called Command is in this project, but command is not a Command object. 

Maybe change it to `commandString`?

Smanmos commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,107 @@
+ */
+public class Ui {
+    Scanner sc;
+    private static TaskList taskList;
+    private static Storage storage;
```
Why are taskList and storage static?

Smanmos commented on src/main/java/duke/command/deadlineCommand.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents the user command when the user enters a deadline.
+ */
+public class deadlineCommand extends Command {
```
Class name should be PascalCase: `DeadlineCommand`

Smanmos commented on src/main/java/duke/command/InvalidCommand.java:
```diff
@@ -0,0 +1,45 @@
+     * @param taskList TaskList that stores the tasks.
+     * @param storage Storage that deals with loading tasks from the file and saving tasks in the file.            
+     */
+    public void execute(TaskList taskList, Storage storage) {
```
Should have `@Override` annotation

Smanmos commented on src/main/java/duke/command/deleteCommand.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Represents the user command when the user is deletes a task.
+ */
+public class deleteCommand extends Command {
```
Class name should be PascalCase: `DeleteCommand`

Smanmos commented on src/main/java/duke/command/doneCommand.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Represents the user command when the user is done with a task.
+ */
+public class doneCommand extends Command {
```
Class name should be PascalCase: `DoneCommand`

Smanmos commented on src/main/java/duke/command/eventCommand.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents the user command when the user enters an event.
+ */
+public class eventCommand extends Command {
```
Class name should be PascalCase: `EventCommand`

Smanmos commented on src/main/java/duke/command/findCommand.java:
```diff
@@ -0,0 +1,50 @@
+
+import duke.Ui;
+
+public class findCommand extends Command {
```
Class name should be PascalCase: `FindCommand`

Smanmos commented on src/main/java/duke/command/findCommand.java:
```diff
@@ -0,0 +1,50 @@
+        Ui.printMessage("Here are the matching tasks in your list:");
+        String[] parts = this.command.split(" ", 2);
+        String wordToFind = parts[1];
+        int count = 1;
```
count should start at 0 instead of 1, to make things more intuitive

Smanmos commented on src/main/java/duke/command/listCommand.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Represents the user command when the user wants to view the task list.
+ */
+public class listCommand extends Command {
```
Class name should be PascalCase: `ListCommand`

Smanmos commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+ * An abstract class that represents the various types of commands that can be entered by the user.
+ */
+public abstract class Command {
+    private String command;
```
Perhaps consider making `command` `protected` instead of `private`? It would make the code much cleaner since subclasses make use of this variable a lot

Smanmos reviewed (commented):
There were some violations of code (in particular class name capitalization). There was also a questionable design decision of writing to the file every time a new task was inserted or deleted, which can case performance issues. 

But overall, it's a good program design.

e0661424 committed:
Add branch-Level-10

e0661424 committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10 into Master

e0661424 committed:
Finish Level-10

e0661424 committed:
Enhance exception handling

e0661424 committed:
Add assertions to document important assumptions in the code

None of the existing classes have assertions to detect the source of
logical flaws in the code.

The addition of assertions will increase the probability of identifying
bugs in the program that are caused by violation of certain assumptions
in the program logic.

Let's add assertions to the relevant parts of the code to define
certain assumptions about the program state.

Using assertions is greatly advantageous as assertions are easy to
implement and help document and verify key assumptions that should
hold at various points in the code.

pranav-ganesh committed:
Merge branch-A-Assertions into master

Add assertions to document important assumptions in the code

e0661424 committed:
Refactor code to improve code quality

There are several instances in the code where code quality can be
further enhanced.

Good quality of code makes sure that code is written in such a way
that makes them highly readable by a human.

Let's improve the code quality by critically examining the code and
refactoring parts of it to make it more readable and make sure it
complies with the code quality commonly practiced guidelines.

Doing it this way will help eliminate common code quality problems
such as weak SLAP, arrow-head style code, too-long methods and
too-deep nesting.

pranav-ganesh committed:
Merge branch-A-CodeQuality into master

e0661424 committed:
Add C-Update

e0661424 committed:
Add Ui.png and enhance GUI

e0661424 committed:
Add User Guide

e0661424 committed:
Add Javadoc to more files

e0661424 committed:
Add final touches to the code

e0661424 committed:
Update User Guide

e0661424 committed:
Further Update User Guide
